substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/SecuritySubstitutions.java

```diff
@@ -77,102 +79,108 @@
     }
 
-    @Substitute
-    private static <T> T doPrivilegedWithCombiner(PrivilegedExceptionAction<T> action, AccessControlContext context, Permission... perms) throws Throwable {
+}
+
+@InternalVMMethod
+class AccessControllerUtil {
+
+    static final AccessControlContext NO_CONTEXT_SINGLETON;
+    static final ThreadLocal<Deque<AccessControlContext>> accStack = ThreadLocal.withInitial(LinkedList::new);
+
+    static {
         try {
-            return action.run();
-        } catch (Throwable ex) {
-            throw AccessControllerUtil.wrapCheckedException(ex);
+            NO_CONTEXT_SINGLETON = ReflectionUtil.lookupConstructor(AccessControlContext.class, ProtectionDomain[].class, boolean.class).newInstance(new ProtectionDomain[0], true);
+        } catch (ReflectiveOperationException ex) {
+            throw VMError.shouldNotReachHere(ex);
         }
     }
 
-    @Substitute
-    private static <T> T doPrivileged(PrivilegedExceptionAction<T> action, AccessControlContext context) throws Throwable {
-        try {
-            return action.run();
-        } catch (Throwable ex) {
-            throw AccessControllerUtil.wrapCheckedException(ex);
+    /* From JDK15's AccessController */
+    static AccessControlContext checkContext(AccessControlContext context,
+                    Class<?> caller) {
+        // check if caller is authorized to create context
+        if (System.getSecurityManager() != null &&
+                        context != null && !SubstrateUtil.cast(context, Target_java_security_AccessControlContext.class).isAuthorized &&
+                        context != NO_CONTEXT_SINGLETON) {
+            ProtectionDomain callerPD = caller.getProtectionDomain();
+            if (callerPD != null && !callerPD.implies(SecurityConstants.CREATE_ACC_PERMISSION)) {
+                return NO_CONTEXT_SINGLETON;
+            }
         }
+        return context;
     }
 
-    @Substitute
-    private static void checkPermission(Permission perm) throws AccessControlException {
-    }
+    /* From JDK15's AccessController */
+    @SuppressWarnings("unused")
+    static <T> T executePrivileged(PrivilegedAction<T> action, AccessControlContext context, Class<?> caller) throws Throwable {
+        if (action == null) {
+            throw new NullPointerException("Null action");
+        }
 
-    @Substitute
-    private static AccessControlContext getContext() {
-        return AccessControllerUtil.NO_CONTEXT_SINGLETON;
-    }
+        accStack.get().push(context);
 
-    @Substitute
-    private static AccessControlContext createWrapper(DomainCombiner combiner, Class<?> caller, AccessControlContext parent, AccessControlContext context, Permission[] perms) {
-        return AccessControllerUtil.NO_CONTEXT_SINGLETON;
+        try {
+            return action.run();
+        } catch (Throwable ex) {
+            throw wrapCheckedException(ex);
+        } finally {
+            if (context != null) {
```

Comment:
pushing the `context` unconditionally but only popping it when `!= null` does not sound correct

Reply:
This was remnant from debugging some failing gate. Should be fixed.